pkg/tms/tmsUtils.go

```diff
@@ -0,0 +1,358 @@
+				break
+			}
+		}
+		if !isNodeFound {
+			wrongNodeNames = append(wrongNodeNames, nodeName)
+		}
+	}
+
+	if mtaVersion != "*" && mtaVersion != mtaYamlMap["version"] {
+		errorMessage += "parameter 'mtaVersion' does not match the MTA version in mta.yaml\n"
+	}
+
+	if len(wrongMtaIdExtDescriptors) > 0 || len(wrongExtDescriptorPaths) > 0 || len(wrongNodeNames) > 0 {
+		if len(wrongMtaIdExtDescriptors) > 0 {
+			sort.Strings(wrongMtaIdExtDescriptors)
+			errorMessage += fmt.Sprintf("parameter 'extends' in MTA extension descriptor files %v is not the same as MTA ID or is missing at all\n", wrongMtaIdExtDescriptors)
+		}
+		if len(wrongExtDescriptorPaths) > 0 {
+			sort.Strings(wrongExtDescriptorPaths)
+			errorMessage += fmt.Sprintf("MTA extension descriptor files %v do not exist\n", wrongExtDescriptorPaths)
+		}
+		if len(wrongNodeNames) > 0 {
+			sort.Strings(wrongNodeNames)
+			errorMessage += fmt.Sprintf("nodes %v do not exist. Please check node names provided in 'nodeExtDescriptorMapping' parameter or create these nodes\n", wrongNodeNames)
+		}
+	}
+
+	if errorMessage == "" {
+		return nodeIdExtDescriptorMapping, nil
+	} else {
+		return nil, errors.New(errorMessage)
+	}
+}
+
+func GetYamlAsMap(utils TmsUtils, yamlPath string) (map[string]interface{}, error) {
+	var result map[string]interface{}
+	bytes, err := utils.FileRead(yamlPath)
+	if err != nil {
+		return result, err
+	}
+	err = yaml.Unmarshal(bytes, &result)
+	if err != nil {
+		return result, err
+	}
+
+	return result, nil
+}
+
+func SetupCommunication(config Options) (communicationInstance CommunicationInterface) {
+	client := &piperHttp.Client{}
+	proxy := config.Proxy
+	options := piperHttp.ClientOptions{}
+	if proxy != "" {
+		transportProxy, err := url.Parse(proxy)
+		if err != nil {
+			log.Entry().WithError(err).Fatalf("Failed to parse proxy string %v into a URL structure", proxy)
+		}
+
+		options = piperHttp.ClientOptions{TransportProxy: transportProxy}
+		client.SetOptions(options)
+		if config.Verbose {
+			log.Entry().Infof("HTTP client instructed to use %v proxy", proxy)
+		}
+	}
+
+	serviceKey, err := unmarshalServiceKey(config.TmsServiceKey)
+	if err != nil {
+		log.Entry().WithError(err).Fatal("Failed to unmarshal TMS service key")
+	}
+	log.RegisterSecret(serviceKey.Uaa.ClientSecret)
+
+	if config.Verbose {
+		log.Entry().Info("Will be used for communication:")
+		log.Entry().Infof("- client id: %v", serviceKey.Uaa.ClientId)
+		log.Entry().Infof("- TMS URL: %v", serviceKey.Uri)
+		log.Entry().Infof("- UAA URL: %v", serviceKey.Uaa.Url)
+	}
+
+	commuInstance, err := NewCommunicationInstance(client, serviceKey.Uri, serviceKey.Uaa.Url, serviceKey.Uaa.ClientId, serviceKey.Uaa.ClientSecret, config.Verbose, options)
+	if err != nil {
+		log.Entry().WithError(err).Fatal("Failed to prepare client for talking with TMS")
+	}
+	return commuInstance
+}
+
+func UploadDescriptors(config Options, communicationInstance CommunicationInterface, utils TmsUtils) error {
```

Comment:
This function is quite complex. Is there a way to split it? For example by moving the for each `nodeIdExtDescriptorMapping` loop out. Don't know if it improves things much, just an idea.